adafruit_ble/hci.py

```diff
@@ -0,0 +1,114 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2020 Dan Halbert for Adafruit Industries
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+Sets up an on-board HCI BLE adapter.
+"""
+
+import time
+
+import board
+import busio
+import digitalio
+import _bleio
+
+
+def create_adapter_esp32_hci(
```

Comment:
Why is this a function instead of a subclass or wrapper of Adapter?

Why does it construct DigitalInOuts itself? Doing so prevents its used with other DigitalInOut-like objects. (Maybe we need to go the other way and provide native pins to Adapter to prevent this.)

Reply:
I agree this could be in a subclass of `Adapter`. That's probably easier than wrapping, since `Adapter` has a wide API.

The original idea was to keep this code out of the main BLE library so it didn't need to be imported unless necessary. But it could be `hci_adapter.py` which implements `HCIAdapter`.

This is a completely hardware specific initialization routine, for ESP32 only. Since it needs to toggle pins to do a hard reset, I created `DigitalInOut`s for this. One of these is shared, so it gets passed on.

Also see https://github.com/adafruit/circuitpython/pull/3310#pullrequestreview-479015091

Reply:
I think this should be a separate library, Adafruit_CircuitPython_ESP32SPI_BLE or something. It isn't generically needed for BLE.

Reply:
After our discussion, created https://github.com/adafruit/Adafruit_CircuitPython_AirLift library, now used by this code.